src/inverse.jl

```diff
@@ -0,0 +1,176 @@
+
+See also: [`inverse`](@ref), [`right_inverse`](@ref), [`@register_inverse`](@ref).
+"""
+function left_inverse end
+
+"""
+    right_inverse(f)
+
+Given a single-input single-output function `f`, return its right inverse `g`. This
+requires that `f` is surjective. If `right_inverse` is defined for a function,
+`left_inverse` and `inverse` must not be defined and should error. `left_inverse(g)`
+should also be defined to return `f`.
+
+See also [`inverse`](@ref), [`left_inverse`](@ref), [`@register_inverse`](@ref).
+"""
+function right_inverse end
+
+"""
+    @register_inverse f g
+    @register_inverse f g left
+    @register_inverse f g right
+
+Mark `f` and `g` as inverses of each other. By default, assume that `f` and `g` are
+bijective. Also defines `left_inverse` and `right_inverse` to call `inverse`. If the
+third argument is `left`, assume that `f` is injective and `g` is its left inverse. If
+the third argument is `right`, assume that `f` is surjective and `g` is its right
+inverse.
+"""
+macro register_inverse(f, g, dir::QuoteNode = :(:both))
+    dir = dir.value
+    if dir == :both
+        quote
+            (::typeof($inverse))(::typeof($f)) = $g
+            (::typeof($inverse))(::typeof($g)) = $f
+            (::typeof($left_inverse))(::typeof($f)) = $(inverse)($f)
+            (::typeof($right_inverse))(::typeof($f)) = $(inverse)($f)
+            (::typeof($left_inverse))(::typeof($g)) = $(inverse)($g)
+            (::typeof($right_inverse))(::typeof($g)) = $(inverse)($g)
+        end
+    elseif dir == :left
+        quote
+            (::typeof($left_inverse))(::typeof($f)) = $g
+            (::typeof($right_inverse))(::typeof($g)) = $f
+        end
+    elseif dir == :right
+        quote
+            (::typeof($right_inverse))(::typeof($f)) = $g
+            (::typeof($left_inverse))(::typeof($g)) = $f
+        end
+    else
+        throw(ArgumentError("The third argument to `@register_inverse` must be `left` or `right`"))
+    end
+end
+
+"""
+    $(TYPEDSIGNATURES)
+
+Check if the provided function has an inverse defined via [`inverse`](@ref). Uses
+`hasmethod` to perform the check.
+"""
+has_inverse(::T) where {T} = hasmethod(inverse, Tuple{T})
+
+"""
+    $(TYPEDSIGNATURES)
+
+Check if the provided function has a left inverse defined via [`left_inverse`](@ref)
+Uses `hasmethod` to perform the check.
+"""
+has_left_inverse(::T) where {T} = hasmethod(left_inverse, Tuple{T})
+
+"""
+    $(TYPEDSIGNATURES)
+
+Check if the provided function has a left inverse defined via [`left_inverse`](@ref)
+Uses `hasmethod` to perform the check.
+"""
+has_right_inverse(::T) where {T} = hasmethod(right_inverse, Tuple{T})
+
+"""
+    $(TYPEDSIGNATURES)
+
+A simple utility function which returns the square of the input. Used to define
+the inverse of `sqrt`.
+"""
+square(x) = x ^ 2
+
+"""
+    $(TYPEDSIGNATURES)
+
+A simple utility function which returns the cube of the input. Used to define
+the inverse of `cbrt`.
+"""
+cube(x) = x ^ 3
```

Comment:
I still think we should lowering sqrt to Pow

Reply:
In that case we wouldn't need to, since the function would not be in the representation.